### PR TITLE
Removed inappropriate pending install item when accessing NVDA menu through systray

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:
None.
Issue [reported](https://groups.io/g/nvda-fr/topics) on a French mailing list.
### Summary of the issue:

When accessing NVDA menu via the systray (enter or right click on the NVDA icon), the item "Install pending update" is present even if there is nothing to install.

In the code, it appears that the function to evaluate whether this item should be present or not is called only when accessing NVDA menu via NVDA+N shortcut, not when accessing it via the systray icon.

### Description of user facing changes

The item "Install pending update" will not be present anymore in NVDA menu when there is no pending update available.<

### Description of development approach
Moved the corresponding piece of code so that it is called in both scenarios.

Note: I have not written deprecation code since 2023.1 is an API breaking release and since this function is probably not used by so many add-ons (if any). I feel that in the case of functions/methods rarely used by add-on authors it's not worth making NVDA's core code heavier with deprecation functions when not necessary. If NVAccess wishes to maintain a deprecated method `evaluateUpdatePendingUpdateMenuItemCommand` in MainFrame

### Testing strategy:
Manual tests with a portable NVDA created from appVeyor build of this PR:
1. Launch NVDA
   * Open NVDA menu via NVDA+N and check that there is no pending install item in the menu
   * Open NVDA menu via systray icon and check that there is no pending install item in the menu
2. Check for update via NVDA menu, download the available (alpha) update but do not install it
   * Open NVDA menu via NVDA+N and check that there is a pending install item in the menu
   * Open NVDA menu via systray icon and check that there is a pending install item in the menu
3. In the `userConfig\updates` folder, delete the .exe file:
   * Open NVDA menu via NVDA+N and check that there is no pending install item in the menu
   * Open NVDA menu via systray icon and check that there is no pending install item in the menu

### Known issues with pull request:
None
### Change log entries:

Bug fixes
`When accessing NVDA menu via its icon in the notification area, NVDA will not suggest a pending update anymore when none is available.`

Changes for developers:
`gui.MainFrame.evaluateUpdatePendingUpdateMenuItemCommand` has been removed; use `gui.MainFrame.sysTrayIcon.evaluateUpdatePendingUpdateMenuItemCommand` instead.

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

File: source/gui/__init__.py [hunk 1]
```diff
@@ -524,8 +514,17 @@ def __init__(self, frame: MainFrame):
 
 		self.Bind(wx.adv.EVT_TASKBAR_LEFT_DOWN, self.onActivate)
 		self.Bind(wx.adv.EVT_TASKBAR_RIGHT_DOWN, self.onActivate)
-
+	
+	def evaluateUpdatePendingUpdateMenuItemCommand(self):
```
Comment (seanbudd) [inline source/gui/__init__.py:518]: If this is something we don't want or expect add-on authors to use, it should be prefixed with an underscore

```suggestion
	def _evaluateUpdatePendingUpdateMenuItemCommand(self):
```

File: source/gui/__init__.py [hunk 2]
```diff
@@ -524,8 +514,17 @@ def __init__(self, frame: MainFrame):
 
 		self.Bind(wx.adv.EVT_TASKBAR_LEFT_DOWN, self.onActivate)
 		self.Bind(wx.adv.EVT_TASKBAR_RIGHT_DOWN, self.onActivate)
-
+	
+	def evaluateUpdatePendingUpdateMenuItemCommand(self):
+		try:
+			self.menu.Remove(self.installPendingUpdateMenuItem)
+		except Exception:
+			log.debug("Error while removing  pending update menu item", exc_info=True)
```
Comment (seanbudd) [inline source/gui/__init__.py:522]: Fix double space in log message

```suggestion
			log.debug("Error while removing pending update menu item", exc_info=True)
```

File: source/gui/__init__.py (hunk 1 above)
Comment (seanbudd) [reply source/gui/__init__.py:518]: Otherwise, we should use deprecations and retain the alias. The current policy is to retain aliases where possible.

Comment (CyrilleB79) [reply source/gui/__init__.py:518]: @seanbudd, it's not very clear to me. Let me rephrase the issues:

There are two `evaluateUpdatePendingUpdateMenuItemCommand` functions (actually, it's a move):
* `gui.MainFrame.evaluateUpdatePendingUpdateMenuItemCommand` that was in NVDA code base and that I remove
* `gui.MainFrame.sysTrayIcon.evaluateUpdatePendingUpdateMenuItemCommand` that was created in this PR.

Questions:
1. In NVDA code base, I do not need gui.MainFrame.evaluateUpdatePendingUpdateMenuItemCommand anymore. Do you recommend to deprecate it or not; it does not cause any problem to keep it; deprecate it would just be useful if we want to clen up the code in the future.
2. I have created gui.MainFrame.sysTrayIcon.evaluateUpdatePendingUpdateMenuItemCommand. Do you recommend to make it private instead?

Comment (seanbudd) [reply source/gui/__init__.py:518]: It's a situation of one or the other.
If this is something we want to remove from the API, as it is not expected to be used by API consumers and/or it is not stable, it should be made private.

Otherwise, it should be deprecated and the alias should remain.
I think in this case, keeping the alias is more appropriate as the code should be stable.

File: source/gui/__init__.py [hunk 3]
```diff
@@ -203,14 +202,8 @@ def onExecuteUpdateCommand(self, evt):
...
 	def evaluateUpdatePendingUpdateMenuItemCommand(self):
-		try:
-			self.sysTrayIcon.menu.Remove(self.sysTrayIcon.installPendingUpdateMenuItem)
-		except:
-			log.debug("Error while removing  pending update menu item", exc_info=True)
-			pass
-		if not globalVars.appArgs.secure and updateCheck and updateCheck.isPendingUpdate():
-			self.sysTrayIcon.menu.Insert(self.sysTrayIcon.installPendingUpdateMenuItemPos,self.sysTrayIcon.installPendingUpdateMenuItem)
-
+		self.sysTrayIcon.evaluateUpdatePendingUpdateMenuItemCommand()
```
Comment (seanbudd) [inline source/gui/__init__.py:205]: can a deprecation warning be added here? once that is done this PR should be fine to be merged.

```suggestion
		log.warning(
			"MainFrame.evaluateUpdatePendingUpdateMenuItemCommand is deprecated. "
			"Use SysTrayIcon.evaluateUpdatePendingUpdateMenuItemCommand instead.",
			stack_info=True,
		)
		self.sysTrayIcon.evaluateUpdatePendingUpdateMenuItemCommand()
		
```

File: source/gui/__init__.py (hunk 1 above)
Comment (seanbudd) [reply source/gui/__init__.py:521]: These are mutually exclusive approaches based on our API policy.

If you wanted to make `gui.SysTrayIcon.evaluateUpdatePendingUpdateMenuItemCommand` private, you would make it as an API breaking change and not retain `gui.MainFrame.evaluateUpdatePendingUpdateMenuItemCommand` as an alias.

Otherwise, you retain the alias with a deprecation notice.
I suggest making change 1, with a deprecation notice.

File: source/gui/__init__.py (hunk 3 above)
Comment (CyrilleB79) [reply source/gui/__init__.py:210]: @seanbudd,, I have applied your suggestion in ad1f2ad4be8a584af6d76db118074fc78eef6d87. This suggestion is in line with what is done in the rest of NVDA code base. The PR is ready.

Side note: I do not fully understand how deprecation is managed in NVDA code base. When deprecation applies to a module level variable/function, we have the warning only if `NVDAState._allowDeprecatedAPI` returns True. For methods in a class, we get the warning in any case and we do not check `NVDAState._allowDeprecatedAPI`'s return value. Why?

Comment (seanbudd) [reply source/gui/__init__.py:210]: thanks.

It is not totally clear on how to handle this for class methods and variables.
A wrapper like this might be possible: https://stackoverflow.com/a/30253848/8030743
We just haven't scheduledtime to investigate a standard syntax to be used for class variables.

Another alternative is similar to existing code:

```py
class Foo
  if NVDAState._allowDeprecatedAPI:
    def example(): pass
```

**Review by seanbudd — APPROVED** [review]
